### DISCUSSION TITLE
Using SPI on Linux/Darwin target

### DISCUSSION THREAD

**Discussion by botoxparty** [modm-io/modm#1119] [Ideas]
Hey,

I have some USB SPI devices, see https://spidriver.com/ or https://www.amazon.de/-/en/DSD-TECH-SH-U05A-UART-Converter/dp/B0BM9JQZB7/

How would I go about using these SPI interfaces with modm? I'm a bit lost.

Would appreciate any help.

**Comment by rleh** [comment]
Hi,
I'm not sure if I understand correctly what you want to do, but I assume you want to use these USB-SPI adapters to have a SPI master on `hosted-linux`/`hosted-darwin` to use (existing) modm drivers to control SPI peripherals (like sensors/displays)?

Therefor you would need to implement the [`SpiMaster` interface](https://github.com/modm-io/modm/blob/a771042eb5d503acdae5ac18d08c61932b7c3618/src/modm/architecture/interface/spi_master.hpp) in a SPI _platform driver_ for the `hosted-linux`/`hosted-darwin` target (simlar to [e.g. this one for STM32](https://github.com/modm-io/modm/tree/a771042eb5d503acdae5ac18d08c61932b7c3618/src/modm/platform/spi/stm32)).
Looking at [their C driver](https://github.com/jamesbowman/spidriver/blob/9f3a7e75d55eea425eae03529680e6f7302cc042/c/common/spidriver.c) this should be possible.

I might be helpful to have a look at the [`modm:platform:canusb` driver](https://github.com/modm-io/modm/tree/a771042eb5d503acdae5ac18d08c61932b7c3618/src/modm/platform/can/canusb), that does something similar but for a USB-CAN adapter for `hosted-linux` target.

**Comment by botoxparty** [comment]
Yeah you are understanding me perfectly :) 

Sounds like a challenge, I'll see how far i can get with it.

**Comment by botoxparty** [comment]
So I tried but I unfortunately don't have the in-depth knowledge of developing linux drivers to be able to refactor the existing linux driver to be compatible with the project.

If anyone would support me with developing this I would love to work together.
